File: ts/step6-first-class-collection/cars.ts
// cars.ts
import { Car } from './car';

export class Cars {
    private readonly cars: Car[];

    constructor(cars: Car[]) {
        if (cars.length === 0) {
            throw new Error('자동차는 최소 1대 이상이어야 합니다.');
        }
        this.cars = cars;
    }

    public moveAll(randomValues: number[]): void {
        if (randomValues.length !== this.cars.length) {
            throw new Error('랜덤 값의 개수가 자동차 수와 일치해야 합니다.');
        }

        this.cars.forEach((car, index) => {
            car.move(randomValues[index]);
        });
    }

    public findWinners(): string[] {
        const maxPosition = Math.max(...this.cars.map(c => c.getPosition()));

        return this.cars
            .filter(c => c.getPosition() === maxPosition)
            .map(c => c.getName());
    }

    public getCars(): readonly Car[] {
        return this.cars;
    }
}

File: java/step6-first-class-collection/src/main/java/badcode/Cars.java
package badcode;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Cars {

    private final List<Car> cars;

    private Cars(List<Car> cars) {
        validate(cars);
        this.cars = new ArrayList<>(cars);
    }

    public static Cars of(List<Car> cars) {
        return new Cars(cars);
    }

    public static Cars from(List<String> names) {
        return new Cars(names.stream().map(Car::of).toList());
    }

    private void validate(List<Car> cars) {
        if (cars == null || cars.isEmpty()) {
            throw new IllegalArgumentException();
        }
    }

    public List<Car> getCars() {
        return Collections.unmodifiableList(cars);
    }

    public void moveAll(int[] randomValues) {
        validateRandomValues(randomValues);
        for (int i = 0; i < cars.size(); i++) {
            this.cars.get(i).move(randomValues[i]);
        }
    }

    public List<String> findWinners() {
        int maxPosition = 0;
        for (Car car : cars) {
            if (car.getPosition() > maxPosition) {
                maxPosition = car.getPosition();
            }
        }
        List<String> winners = new ArrayList<>();
        for (Car car : cars) {
            if (car.getPosition() == maxPosition) {
                winners.add(car.getName());
            }
        }
        return Collections.unmodifiableList(winners);
    }

    private void validateRandomValues(int[] randomValues) {
        if (randomValues == null || randomValues.length == 0) {
            throw new IllegalArgumentException("랜덤 수 배열은 비어있을 수 없습니다.");
        }
        if (randomValues.length < this.cars.size()) {
            throw new IllegalArgumentException("랜덤 수 배열의 길이는 차 대수보다 작을 수 없습니다.");
        }
    }
}
